### PR TITLE
[Serializer] Add ->hasCacheableSupportsMethod() to CacheableSupportsMethodInterface

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Enhances the interface introduced in #27049 to make it possible to dynamically define if "supports" methods are cacheable.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
@teohhanhui we not a new name. Do you have a suggestion? Naming is hard :) About a separate interface, I don't understand what you're suggesting. One interface is enough, isn't it?

**Comment by emodric** [comment]
@nicolas-grekas Apart from the typo( `CacheableSupportsMethodInteface` -> `CacheableSupportsMethodInterface`), this still doesn't work unfortunately. If it means anything, my normalizers do not extend `AbstractNormalizer` and only implement `NormalizerInterface` directly.

What is your aim here? Should I implement this interface in my normalizers and return `false`, because it doesn't work with that too.

**Comment by nicolas-grekas** [comment]
What does "doens't work" mean? My aim it at making the cacheable mechanism BC, by making it opt-in.

**Comment by nicolas-grekas** [comment]
What does "doesn't work" mean? My aim it at making the cacheable mechanism BC, by making it opt-in.

**Comment by emodric** [comment]
It means, my normalizers are still not used and fall back to `ObjectNormalizer`, as per my previous comment on #27049

**Comment by nicolas-grekas** [comment]
@emodric I fail to understand why. Can you debug the situation after applying this PR please?

**Comment by dunglas** [comment]
Can’t we mark `supportsNormalization` methods `@final` instead to avoid the BC break? It means that someone wanting to create a “dynamic” method would have to use decoration instead of inheritance (a good practice anyway).

**Comment by emodric** [comment]
I'll try. Give me a couple of days.

**Comment by teohhanhui** [comment]
I think there's a much simpler and BC-safe way of doing this. Have subclasses of the built-in normalizers that implement the new interface. Then it's truly opt-in and has no potential of any BC breaks.

**Comment by teohhanhui** [comment]
And we could of course use them by default in the declared services.

**Comment by dunglas** [comment]
@teohhanhui we had this idea, but it only fixes the problem when using the component directly. `FrameworkBundle` register the existing classes as services, and replacing them by the subclasses would be another BC break...

**Comment by dunglas** [comment]
@emodric can you provide the code of your custom normalizer (even privately, using a private Gist for instance), or PM on Slack? Does it extend a builtin normalizer, or the `abstract` class directly?

**Comment by emodric** [comment]
I'm not near a computer today, but as soon as I'm back, I'll do it. I'll try to create a test app to provoke the issue, too.

In the meantime, my normalizers implement `NormalizerInterface` directly, without using `AbstractNormalizer` or extending any other normalizer. They are tagged with `serializer.normalizer` with default priority.

`supportsNormalization` dynamically checks for supported data by checking instance of provided data as well as a property in the provided object.

**Comment by teohhanhui** [comment]
@dunglas:
> FrameworkBundle register the existing classes as services, and replacing them by the subclasses would be another BC break...

How so? Liskov Substitution Principle means we can substitute them with any subclasses any time, no?

**Comment by nicolas-grekas** [comment]
> Then it's truly opt-in and has no potential of any BC breaks.

@teohhanhui what is not truly opt-in in the current PR? Also, where is the potential BC break? Please advise I don't understand. Also, how would you handle the ArrayNormalizer case (see attached patch for the way this PR does it?) Would you mind opening a PR embedding your proposal? It might be easier to understand each other this way. Thanks.

**Comment by dunglas** [comment]
> Liskov Substitution Principle means we can substitute them with any subclasses any time, no?

Unfortunately we cannot rely on it to replace Symfony native services. If the user use `get_class($service) === 'Foo\Bar'` in his code (instead of using `instanceof`), it will break... And it's very frequent.

**Review by dunglas — APPROVED** [review]

**Review by sroze — APPROVED** [review]
👍 (I understand the reluctance for such thing but don't see any better way to solve the problem)

**Comment by emodric** [comment]
@dunglas Here's a Symfony Standard app with @nicolas-grekas `ser-cache` branch configured, with a sample normalizer which shows the bug: https://github.com/emodric/symfony-serializer-bug

Run the app with the built in server with `php bin/console server:run -d web` and access the homepage.

The output is `{"fooBar": "foo", "fooBaz": "bar"}`, while I'd expect it to be ``{"foo_bar": "foo", "foo_baz": "bar"}`` showing that sample normalizer is not being used, instead the fallback to `ObjectNormalizer` is being done.

**Comment by nicolas-grekas** [comment]
@emodric thanks for the reproducer, it definitely helped, this is now fixed!

File: src/Symfony/Component/Serializer/Serializer.php [hunk 1]
```diff
@@ -221,12 +221,11 @@ private function getNormalizer($data, ?string $format, array $context)
...
                 }
 
-                if (!$normalizer instanceof CacheableSupportsMethodInterface) {
+                if (!$normalizer instanceof CacheableSupportsMethodInterface || !$normalizer->hasCacheableSupportsMethod()) {
                     $this->normalizerCache[$format][$type][$k] = false;
                 } elseif ($normalizer->supportsNormalization($data, $format)) {
                     $this->normalizerCache[$format][$type][$k] = true;
-
-                    return $normalizer;
+                    break;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Serializer.php]: this was the bug spotted by @emodric: we cannot return here, as the non-cacheable normalizer weren't  called yet.

**Review by teohhanhui — APPROVED** [review]

**Comment by emodric** [comment]
Thanks @nicolas-grekas ! Tested it with my test suite and now it works :+1:

**Comment by teohhanhui** [comment]
What about:

VarySupportsNormalizerInterface
- varySupportsNormalization
- varySupportsDenormalization

(So that there's more granular control and we can cache separately for normalization / denormalization?)

VarySupportsSerializerInterface
- varySupports

**Comment by dunglas** [comment]
Actually, we need to add different interfaces for Normalizers and Denormalizers.

**Comment by teohhanhui** [comment]
The name `CacheableSupportsMethodInterface` is misleading if it can be either cacheable or not cacheable. (When the name says "cacheable", it better mean just that.)

**Pull request merged**